### PR TITLE
[Text Analytics] Add support for dynamic classification

### LINKED ISSUE #31271: [Text Analytics] Add support for dynamic classification
Reference: https://github.com/Azure/azure-rest-api-specs/blob/main/dev/cognitiveservices/data-plane/Language/analyzetext.json

# Tasks
- [ ] Implementation
- [ ] Tests
- [ ] Docs

### PR DESCRIPTION
**Pull request by joseharriaga**
Resolves: #31271

Features added:
- Added the following methods: `TextAnalyticsClient.DynamicClassify` and `TextAnalyticsClient.DynamicClassifyAsync`.
- Added the following methods: `TextAnalyticsClient.DynamicClassifyBatch` and `TextAnalyticsClient.DynamicClassifyBatchAsync`.
- Added the following classes: `DynamicClassifyDocumentResultCollection` and `DynamicClassifyOptions`.
- Added the `ClassificationType` enum.

### CODE REVIEW

File: sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsModelFactory.cs
Comment (kinelski) [inline sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsModelFactory.cs:740]: Just double checking, but you intend to add factory methods in a future PR. Is that right?

File: sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClient.cs
Comment (joseharriaga) [inline sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClient.cs:2127]: @heaths, @kinelski: I wanted to ask you what you think about this one: Right now, `DynamicClassifyOptions` has the required `Categories` property, which is just a list of strings to specify which categories participate in the classification. Here's their example, where they use the categories "Health", "Politics", "Music", and "Sports":
🔗 https://github.com/Azure/azure-rest-api-specs/blob/main/specification/cognitiveservices/data-plane/Language/preview/2022-10-01-preview/examples/text/SuccessfuDynamicClassificationRequest.json

From the Guidelines: "The Options class is designed similarly to .NET custom attributes, where required service method parameters are modeled as Options class constructor parameters and get-only properties, and optional parameters are get-set properties." 

(I just realized I need to make `Categories` get-only if I wanted to follow this accurately. But before I do that...)

I was wondering if in this case it makes more sense to move the `Categories` property out of the Options class and make it a parameter that must be passed in all these methods (kind of like `document`/`documents` is its own parameter).

File: sdk/textanalytics/Azure.AI.TextAnalytics/api/Azure.AI.TextAnalytics.netstandard2.0.cs
Comment (joseharriaga) [reply sdk/textanalytics/Azure.AI.TextAnalytics/api/Azure.AI.TextAnalytics.netstandard2.0.cs:453]: I think it's similar but not entirely the same: My understanding is that `CategoriesFilter` in `RecognizePiiEntitiesOptions` is an optional property that filters down the result; is this correct? On the other hand, `Categories` in `DynamicClassifyOptions` is a required property that specifies the entire set of categories that will be used during the classification.

File: sdk/textanalytics/Azure.AI.TextAnalytics/api/Azure.AI.TextAnalytics.netstandard2.0.cs
Comment (joseharriaga) [reply sdk/textanalytics/Azure.AI.TextAnalytics/api/Azure.AI.TextAnalytics.netstandard2.0.cs:452]: I passed it as a constructor parameter because `Categories` is a required property, but see my following comment because I think there's a better way and want to know what you think: https://github.com/Azure/azure-sdk-for-net/pull/32593#discussion_r1028375249

**Review by kristapratico — APPROVED** [review]
Dynamic classification changes LGTM. Just one question - since you share ClassifyDocumentResult between DC and custom single/multi label classification, do you plan to put the detectedLanguage on this result type (even though DC can't do auto-detect language)? I don't recall if .NET decided to separate result types which can't do auto-detect or not.

File: sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsModelFactory.cs
Comment (joseharriaga) [reply sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsModelFactory.cs]: Oh, funny. I suspect I lost this change because I might have forgotten to save the file before I rebased my branch. 🤔 I have added it back. Thank you for catching it!

**Review by mssfang — APPROVED** [review]

File: sdk/textanalytics/Azure.AI.TextAnalytics/api/Azure.AI.TextAnalytics.netstandard2.0.cs
Comment (heaths) [reply sdk/textanalytics/Azure.AI.TextAnalytics/api/Azure.AI.TextAnalytics.netstandard2.0.cs:452]: What's not clear is whether the property should be read-only henceforth. We don't want customers clearing it before sending if it's required (by default we should not typically be serializing an empty collection). `ICollection<T>` defines `Clear`, so `IList<T>` would have it to. I would talk with @KrzysztofCwalina and @tg-msft about this. Perhaps if it's passed in as a parameter, backing properties should be read-only which means customers can't mutate it after. If anything, that is equivalent to other constructor parameters now that should be read-only (i.e. no `set` accessor) for those backing constructor parameters.

File: sdk/textanalytics/Azure.AI.TextAnalytics/src/DynamicClassifyOptions.cs
Comment (heaths) [inline sdk/textanalytics/Azure.AI.TextAnalytics/src/DynamicClassifyOptions.cs:30]: 1. As mentioned in one of my first comments, this should probably be `IReadOnlyList<T>` (or `IReadOnlyCollection<T>`, if more consistent throughout). This jives with our guidelines for constructor parameters (required properties).
2. Do not initialize to a `new List<string>` here. Do so in the constructor since you're taking args anyway.. This is a wasted allocation.
3. If this is for a required parameter, `categories` should never be null. Throw an exception. Use `Argument.AssertNotNullOrEmpty` because it's required, which means it has to be non-empty.

File: sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClient.cs
Comment (heaths) [reply sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClient.cs:2127]: Depends if it might grow. I'd talk with @tg-msft or @KrzysztofCwalina about this. There's some nuance to this. I've linked the guidelines above as well but it's better to get a final say from them.

File: sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClient.cs
Comment (heaths) [inline sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClient.cs:2124]: Nit: 2 paragraphs just to document this way seems unnecessary and long. This won't grow. It's more typical just to say something like:

```c#
/// <exception cref="ArgumentNullException"><paramref name="document"/> or <paramref name="options"/> is null.</exception>
```

That also jives with official .NET documentation.

**Review by heaths — CHANGES_REQUESTED** [review]
I have enough concerns around the options class I'd like to re-review when this changes.

File: sdk/textanalytics/Azure.AI.TextAnalytics/api/Azure.AI.TextAnalytics.netstandard2.0.cs
Comment (joseharriaga) [reply sdk/textanalytics/Azure.AI.TextAnalytics/api/Azure.AI.TextAnalytics.netstandard2.0.cs:1581]: The reason why it's optional is because it currently contains a property that is required for the request to be valid. This is why I find this pattern awkward and why I want to move that property out of the Options class and make it a method parameter instead (see my initial comment here: https://github.com/Azure/azure-sdk-for-net/pull/32593#discussion_r1028375249)

File: sdk/textanalytics/Azure.AI.TextAnalytics/api/Azure.AI.TextAnalytics.netstandard2.0.cs
Comment (joseharriaga) [reply sdk/textanalytics/Azure.AI.TextAnalytics/api/Azure.AI.TextAnalytics.netstandard2.0.cs:326]: I have added this one to my notes to discuss it in the future before reviewing APIView. The precedent of "multi label" might need to be considered, but in general, I would prefer "multiple".

File: sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClient.cs
Comment (joseharriaga) [reply sdk/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsClient.cs]: We're going with the option of moving the `Categories` property out of the `DynamicClassifyOptions` class and making it a parameter. I think it looks better now.

**Review by kinelski — APPROVED** [review]
LGTM!

**Review by heaths — APPROVED** [review]
Approving to unblock, but I have some minor questions/concerns.

File: sdk/textanalytics/Azure.AI.TextAnalytics/src/ServiceClients/ServiceClient.cs
Comment (joseharriaga) [reply sdk/textanalytics/Azure.AI.TextAnalytics/src/ServiceClients/ServiceClient.cs:166]: The difference is between the LRO features (e.g. `StartAnalyzeHealthcareEntities`/`StartAnalyzeHealthcareEntitiesAsync`) and the non-LRO features (e.g., `ExtractKeyPhrases`/`ExtractKeyPhrasesAsync` and `ExtractKeyPhrasesBatch`/`ExtractKeyPhrasesBatchAsync`). Dynamic classification is the latter, although I can see it being a little confusing because custom classification is the former.

**Pull request merged**
